pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,381 @@
+}
+
+// OnOutboundCluster implements the Plugin interface method.
+func (Plugin) OnOutboundCluster(env model.Environment, node model.Proxy, service *model.Service,
+	servicePort *model.Port, cluster *xdsapi.Cluster) {
+}
+
+func isRbacEnabled(store model.IstioConfigStore) bool {
+	var configProto *rbacproto.RbacConfig
+	config := store.RbacConfig(RbacConfigName, kube.IstioNamespace)
+	if config != nil {
+		configProto = config.Spec.(*rbacproto.RbacConfig)
+	}
+	if configProto == nil {
+		log.Debugf("rbac plugin disabled: No RbacConfig")
+		return false
+	}
+	// TODO(yangminzhu): Supports ON_WITH_INCLUSION and ON_WITH_EXCLUSION.
+	if configProto.Mode == rbacproto.RbacConfig_ON {
+		return true
+	}
+
+	log.Debugf("rbac plugin disabled by RbacConfig: %v", *configProto)
+	return false
+}
+
+// buildHTTPFilter builds the RBAC http filter that enforces the access control to the specified
+// service which is co-located with the sidecar proxy.
+func buildHTTPFilter(hostName model.Hostname, store model.IstioConfigStore) *http_conn.HttpFilter {
+	namespace := hostName.Namespace()
+	roles := store.ServiceRoles(namespace)
+	if roles == nil {
+		log.Debugf("no service roles found for %s in namespace %s", hostName, namespace)
+		return nil
+	}
+	bindings := store.ServiceRoleBindings(namespace)
+	config := convertRbacRulesToFilterConfig(hostName.String(), roles, bindings)
+	return &http_conn.HttpFilter{
+		Name:   RbacFilterName,
+		Config: util.MessageToStruct(config),
+	}
+}
+
+// convertRbacRulesToFilterConfig converts the current RBAC rules (ServiceRole and ServiceRoleBindings)
+// in service mesh to the corresponding proxy config for the specified service. The generated proxy config
+// will be consumed by envoy RBAC filter to enforce access control on the specified service.
+func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindings []model.Config) *rbacconfig.RBAC {
+	// roleToBinding maps ServiceRole name to a list of ServiceRoleBindings.
+	roleToBinding := map[string][]*rbacproto.ServiceRoleBinding{}
+	for _, binding := range bindings {
+		bindingProto := binding.Spec.(*rbacproto.ServiceRoleBinding)
+		roleName := bindingProto.RoleRef.Name
+		if roleName == "" {
+			log.Errorf("ignored invalid binding with empty RoleRef.Name: %v", *bindingProto)
+			continue
+		}
+		roleToBinding[roleName] = append(roleToBinding[roleName], bindingProto)
+	}
+
+	rbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
+	for _, role := range roles {
+		// Constructs the policy for each ServiceRole.
+		var policy *policyproto.Policy
+		principals := convertToPrincipals(roleToBinding[role.Name])
+		log.Debugf("checking role %v for service %v", role.Name, service)
+		for i, rule := range role.Spec.(*rbacproto.ServiceRole).Rules {
+			//TODO(yangminzhu): Also check the destination related properties.
+			if stringMatch(service, rule.Services) {
```

Comment:
Shall we filter out the roles/bindings that do not apply to "service" first? This avoids unnecessary convertToPrincipals() calls.

Reply:
Good point. I delayed the initialize of principals after we found a match.